File: docs/snippets/all/archetypes/pinhole_simple.rs
//! Log a pinhole and a random image.

use ndarray::{Array, ShapeBuilder as _};
use rand::prelude::*;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_pinhole").spawn()?;

    let mut image = Array::<u8, _>::default((3, 3, 3).f());
    let mut rng = rand::rngs::SmallRng::seed_from_u64(42);
    image.map_inplace(|x| *x = rng.random());

    rec.log(
        "world/image",
        &rerun::Pinhole::from_focal_length_and_resolution([3., 3.], [3., 3.]),
    )?;
    rec.log(
        "world/image",
        &rerun::Image::from_color_model_and_tensor(rerun::ColorModel::RGB, image)?,
    )?;

    Ok(())
}

File: docs/snippets/all/archetypes/pinhole_simple.cpp
// Log a pinhole and a random image.

#include <rerun.hpp>

#include <algorithm> // std::generate
#include <cstdlib>   // std::rand
#include <vector>

int main(int argc, char* argv[]) {
    const auto rec = rerun::RecordingStream("rerun_example_pinhole");
    rec.spawn().exit_on_failure();

    rec.log("world/image", rerun::Pinhole::from_focal_length_and_resolution(3.0f, {3.0f, 3.0f}));

    std::vector<uint8_t> random_data(3 * 3 * 3);
    std::generate(random_data.begin(), random_data.end(), [] {
        return static_cast<uint8_t>(std::rand());
    });

    rec.log("world/image", rerun::Image::from_rgb24(random_data, {3, 3}));
}
